q2cli/tests/test_tools.py

```diff
@@ -15,10 +15,109 @@
+                    'numbers:numerical', '--cast', 'numbers:categorical',
+                    '--output-file', self.output_file])
+
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn(
+            '"numbers" appears in cast more than once.', result.output)
+
+    def test_input_invalid_cast_format(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast', 'numbers',
+                    '--output-file', self.output_file])
+
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn(
+            'Could not parse provided cast arguments into unique COLUMN:TYPE'
+            ' pairs.', result.output)
+
+    def test_error_on_extra(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'extra:numeric', '--output-file', self.output_file])
+
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn(
+            "The following cast columns were not found within the"
+            " metadata: extra", result.output)
+
+    def test_error_on_missing(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'numbers:categorical', '--error-on-missing',
+                    '--output-file', self.output_file])
+
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn(
+            "The following columns within the metadata"
+            " were not provided in the cast: strings",
+            result.output)
+
+    def test_extra_columns_removed(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'numbers:categorical', '--cast', 'extra:numeric',
+                    '--ignore-extra', '--output-file', self.output_file])
+
+        self.assertEqual(result.exit_code, 0)
+        casted_metadata = _load_metadata(self.output_file)
+        self.assertNotIn('extra', casted_metadata.columns.keys())
+
+    def test_complete_successful_run(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'numbers:categorical', '--output-file', self.output_file])
+
+        self.assertEqual(result.exit_code, 0)
+        input_metadata = _load_metadata(self.metadata_file)
+        self.assertEqual('numeric', input_metadata.columns['numbers'].type)
+
+        casted_metadata = _load_metadata(self.output_file)
+        self.assertEqual('categorical',
+                         casted_metadata.columns['numbers'].type)
+
+    def test_write_to_stdout(self):
+        result = self.runner.invoke(
+            tools, ['cast-metadata', self.metadata_file, '--cast',
+                    'numbers:categorical'])
+
+        self.assertEqual(result.exit_code, 0)
+        self.assertEqual(self.cast_metadata_dump, result.output)
+
+
```

Comment:
Let's add one more test to help us out here if any additional column types get added to the `qiime2` framework:

```python
from qiime2.metadata.base import SUPPORTED_COLUMN_TYPES

def test_valid_column_types(self):
    result = self.runner.invoke(tools, ['cast-metadata', '--help'])
    for col_type in SUPPORTED_COLUMN_TYPES:
        self.assertIn(col_type, result.output)
```

This is untested code, so you'll probably need to tweak it a bit, but this should make it so that this test will fail if a new column type is added (like `time`) - prompting us to go ahead and update the hardcoded column types in `tools.py`.

Also note, we don't care about slow import times nearly as much here in the test suite.

Reply:
![Screen Shot 2021-07-21 at 2 41 40 PM](https://user-images.githubusercontent.com/274668/126563917-46004aff-494c-45d2-a64c-efcb0252e1f6.png)

Perfect! That is just what I was looking for!

Reply:
@thermokarst thanks, this is a great addition! Just tested this locally by setting `SUPPORTED_COLUMN_TYPES = {'categorical', 'numeric', 'footype'}` which results in a failure - so this is perfect. Pushed up this added test and those final tweaks to the leftover typos from the cache work, so we should be just about done - let me know if there's anything else we missed!

Reply:
@thermokarst thanks, this is a great addition! Just tested this locally by setting `SUPPORTED_COLUMN_TYPES = {'categorical', 'numeric', 'footype'}` which results in a failure - so this is perfect. Pushed up this added test and those final tweaks to the leftover typos from the cache work, so we should be just about done - let me know if there's anything else we missed!